From 737af3d3aea4e886e391c16e6ae6c4b5d11fec1a Mon Sep 17 00:00:00 2001
From: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
Date: Mon, 6 Apr 2026 09:52:11 +0530
Subject: [PATCH 3/9] config: Add enable_smd parameter config

Add enable_smd configuration parameter (default: 0) to control SMD
feature enablement in both agent and controller. This allows conditional
registration of SMD TLV handlers based on deployment requirements.

- Add to beerocks_agent.conf.in and beerocks_controller.conf.in
- Add to config structures (sConfigSlave, sDbMasterConfig, sDeviceConf)
- Parse and store in agent/controller initialization

Signed-off-by: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
---
 agent/config/beerocks_agent.conf.in                     | 1 +
 agent/src/beerocks/slave/agent_db.h                     | 1 +
 agent/src/beerocks/slave/beerocks_agent_main.cpp        | 1 +
 common/beerocks/bcl/include/bcl/beerocks_config_file.h  | 2 ++
 common/beerocks/bcl/source/beerocks_config_file.cpp     | 2 ++
 controller/config/beerocks_controller.conf.in           | 3 +++
 controller/src/beerocks/master/beerocks_master_main.cpp | 1 +
 controller/src/beerocks/master/db/db.h                  | 1 +
 8 files changed, 12 insertions(+)

diff --git a/agent/config/beerocks_agent.conf.in b/agent/config/beerocks_agent.conf.in
index c158c3c80..1f3a03dcc 100644
--- a/agent/config/beerocks_agent.conf.in
+++ b/agent/config/beerocks_agent.conf.in
@@ -17,6 +17,7 @@ bridge_iface=@BEEROCKS_BRIDGE_IFACE@
 enable_system_hang_test=0 # 0 - disabled
 monitor_polling_rate_msec=@BEEROCKS_MONITOR_POLLING_RATE_MSEC@
 enable_auto_chansel_handling=0 # 0 - disabled
+enable_smd=0 # 0 - disabled, 1 - enabled (SMD - Seamless Mobility Domain)
 
 # This variable is a workaround that turn on the profile_x_disallow flag when configuring the
 # traffic separation, if the Controller sent unsupported configuration on the profile disallow
diff --git a/agent/src/beerocks/slave/agent_db.h b/agent/src/beerocks/slave/agent_db.h
index 0b80e8a6f..1daf24243 100644
--- a/agent/src/beerocks/slave/agent_db.h
+++ b/agent/src/beerocks/slave/agent_db.h
@@ -181,6 +181,7 @@ public:
         uint8_t ap_maximum_links     = 0;
         uint8_t bsta_maximum_links   = 0;
         bool enable_auto_chansel_handling;
+        bool enable_smd                             = false;
         int8_t num_of_bands = 0;
         uint8_t beacon_11k_default_measurement_mode = 2;
 
diff --git a/agent/src/beerocks/slave/beerocks_agent_main.cpp b/agent/src/beerocks/slave/beerocks_agent_main.cpp
index 437c2f00f..4f3177f1b 100644
--- a/agent/src/beerocks/slave/beerocks_agent_main.cpp
+++ b/agent/src/beerocks/slave/beerocks_agent_main.cpp
@@ -560,6 +560,7 @@ static int run_beerocks_slave(beerocks::config_file::sConfigSlave &beerocks_slav
 
         db->device_conf.enable_auto_chansel_handling =
             beerocks_slave_conf.enable_auto_chansel_handling == "1";
+        db->device_conf.enable_smd = beerocks_slave_conf.enable_smd == "1";
         {
             bool uci_enabled = false;
             if (beerocks::bpl::cfg_get_traffic_separation_bridge_mapping_enabled(uci_enabled)) {
diff --git a/common/beerocks/bcl/include/bcl/beerocks_config_file.h b/common/beerocks/bcl/include/bcl/beerocks_config_file.h
index e3fcf5a5f..b611a4b46 100644
--- a/common/beerocks/bcl/include/bcl/beerocks_config_file.h
+++ b/common/beerocks/bcl/include/bcl/beerocks_config_file.h
@@ -40,6 +40,7 @@ public:
         std::string model;
         std::string ucc_listener_port;
         std::string load_legacy_client_roaming;
+        std::string enable_smd;
         std::string load_service_fairness;
         std::string load_backhaul_measurements;
         std::string load_front_measurements;
@@ -96,6 +97,7 @@ public:
         std::string on_boot_scan;
         std::string dwell_time;
         std::string enable_auto_chansel_handling;
+        std::string enable_smd;
         std::string clients_unicast_measurements;
         std::string send_btm_to_non_11v_sta;
         std::string beacon_11k_default_measurement_mode;
diff --git a/common/beerocks/bcl/source/beerocks_config_file.cpp b/common/beerocks/bcl/source/beerocks_config_file.cpp
index f6808e8bc..7994f9176 100644
--- a/common/beerocks/bcl/source/beerocks_config_file.cpp
+++ b/common/beerocks/bcl/source/beerocks_config_file.cpp
@@ -53,6 +53,7 @@ bool config_file::read_master_config_file(const std::string &config_file_path, s
         std::make_tuple("ucc_listener_port=", &conf.ucc_listener_port, mandatory_master),
         std::make_tuple("load_legacy_client_roaming=", &conf.load_legacy_client_roaming,
                         mandatory_master),
+        std::make_tuple("enable_smd=", &conf.enable_smd, 0),
         std::make_tuple("load_service_fairness=", &conf.load_service_fairness, mandatory_master),
         std::make_tuple("load_backhaul_measurements=", &conf.load_backhaul_measurements, 0),
         std::make_tuple("load_front_measurements=", &conf.load_front_measurements, 0),
@@ -148,6 +149,7 @@ bool config_file::read_slave_config_file(const std::string &config_file_path, sC
             std::make_tuple("on_boot_scan=", &conf.on_boot_scan, 0),
             std::make_tuple("dwell_time=", &conf.dwell_time, 0),
             std::make_tuple("enable_auto_chansel_handling=", &conf.enable_auto_chansel_handling, 0),
+            std::make_tuple("enable_smd=", &conf.enable_smd, 0),
             std::make_tuple("clients_unicast_measurements=", &conf.clients_unicast_measurements, 0),
             std::make_tuple("send_btm_to_non_11v_sta=", &conf.send_btm_to_non_11v_sta, 0),
             std::make_tuple("beacon_11k_default_measurement_mode=", &conf.beacon_11k_default_measurement_mode, 0),
diff --git a/controller/config/beerocks_controller.conf.in b/controller/config/beerocks_controller.conf.in
index 89904c35a..63c1faae6 100644
--- a/controller/config/beerocks_controller.conf.in
+++ b/controller/config/beerocks_controller.conf.in
@@ -15,6 +15,9 @@ ucc_listener_port=8002 # 0 - disabled
 #   Client steering feature:
 load_legacy_client_roaming=1
 
+#   SMD feature:
+enable_smd=0 # 0 - disabled, 1 - enabled (SMD - Seamless Mobility Domain)
+
 #   Measurement feature:
 load_backhaul_measurements=0
 load_front_measurements=1
diff --git a/controller/src/beerocks/master/beerocks_master_main.cpp b/controller/src/beerocks/master/beerocks_master_main.cpp
index 9360e810d..a7f8989e7 100644
--- a/controller/src/beerocks/master/beerocks_master_main.cpp
+++ b/controller/src/beerocks/master/beerocks_master_main.cpp
@@ -153,6 +153,7 @@ static void fill_master_config(son::db::sDbMasterConfig &master_conf,
         beerocks::string_utils::stoi(main_master_conf.ucc_listener_port);
     master_conf.load_service_fairness      = (main_master_conf.load_service_fairness == "1");
     master_conf.load_legacy_client_roaming = (main_master_conf.load_legacy_client_roaming == "1");
+    master_conf.enable_smd                 = (main_master_conf.enable_smd == "1");
     master_conf.load_backhaul_measurements = (main_master_conf.load_backhaul_measurements == "1");
     master_conf.load_front_measurements    = (main_master_conf.load_front_measurements == "1");
     master_conf.load_monitor_on_vaps       = (main_master_conf.load_monitor_on_vaps == "1");
diff --git a/controller/src/beerocks/master/db/db.h b/controller/src/beerocks/master/db/db.h
index b1b86b4e4..e03facec9 100644
--- a/controller/src/beerocks/master/db/db.h
+++ b/controller/src/beerocks/master/db/db.h
@@ -122,6 +122,7 @@ public:
         bool load_optimal_path_roaming_prefer_signal_strength;
         bool load_client_11k_roaming;
         bool load_legacy_client_roaming;
+        bool enable_smd;
         bool load_ire_roaming;
         bool load_load_balancing;
         bool load_service_fairness;
-- 
2.34.1

